src/chalk-overview.md

```diff
@@ -0,0 +1,141 @@
+
+The syntax contains things from Rust that we know and love for example traits,
+impls, and struct definitions. Parsing is often the first "phase" of
+transformation that a program goes through in order to become a format that
+chalk can understand.
+
+## Lowering
+
+After parsing, there is a "lowering" phase. This aims to convert traits/impls
+into "program clauses". A [`ProgramClause` (source code)][programclause] is
+essentially one of the following:
+
+* A [clause] of the form `consequence :- conditions` where `:-` is read as
+  "if" and `conditions = cond1 && cond2 && ...`
+* A universally quantified clause of the form
+  `forall<T> { consequence :- conditions }`
+  * `forall<T> { ... }` is used to represent [universal quantification]. See the
+    section on [Lowering to logic][lowering-forall] for more information.
+  * A key thing to note about `forall` is that we don't allow you to "quantify"
+    over traits, only types and regions (lifetimes). That is, you can't make a
+    rule like `forall<Trait> { u32: Trait }` which would say "`u32` implements
+    all traits". You can however say `forall<T> { T: Trait }` meaning "`Trait`
+    is implemented by all types".
+  * `forall<T> { ... }` is represented in the code using the [`Binders<T>`
+    struct][binders-struct].
+
+This is the phase where we encode the rules of the trait system into logic. For
+example, if we have:
+
+```rust,ignore
+impl<T: Clone> Clone for Vec<T> {}
+```
+
+We generate:
+
+```rust,ignore
+forall<T> { (Vec<T>: Clone) :- (T: Clone) }
+```
+
+This rule dictates that `Vec<T>: Clone` is only satisfied if `T: Clone` is also
+satisfied (i.e. "provable").
+
+### Well-formedness checks
+
+As part of lowering from the AST to the internal IR, we also do some "well
+formedness" checks. See the [source code][well-formedness-checks] for where
+those are done. The call to `record_specialization_priorities` checks
+"coherence" which means that it ensures that two impls of the same trait for the
+same type cannot exist.
+
+## Intermediate Representation (IR)
+
+The second intermediate representation in chalk is called, well, the "ir". :)
+The [IR source code][ir-code] contains the complete definition. The
+`ir::Program` struct contains some "rust things" but indexed and accessible in
+a different way. This is sort of analogous to the [HIR] in Rust.
+
+For example, if you have a type like `Foo<Bar>`, we would represent `Foo` as a
+string in the AST but in `ir::Program`, we use numeric indices (`ItemId`).
+
+In addition to `ir::Program` which has "rust-like things", there is also
+`ir::ProgramEnvironment` which is "pure logic". The main field in that is
+`program_clauses` which contains the `ProgramClause`s that we generated
+previously.
+
+## Rules
+
+The `rules` module works by iterating over every trait, impl, etc. and emitting
+the rules that come from each one. The traits section of the rustc-guide (that
```

Comment:
Perhaps link to https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html for convenience?